### PR TITLE
ByteString: fix bugs, type safety, and add performance improvements with tests

### PR DESCRIPTION
**Pull request by pjfanning**
## Summary

This PR fixes several correctness bugs, a type-safety issue, adds performance improvements, and adds targeted test coverage to `ByteString` and `ByteStringBuilder`.

---

### `ByteString1C.writeToBuffer(buffer, offset)` — wrong formula
The formula `Math.min(buffer.remaining, offset + length)` was incorrect: when `offset > 0` it could request more bytes from the backing array than actually exist past the offset, leading to `ArrayIndexOutOfBoundsException`. Fixed to `Math.min(buffer.remaining, length - offset)`.

### `ByteStringBuilder.sizeHint` — `NegativeArraySizeException`
Calling `sizeHint(n)` with `n` smaller than the number of bytes already committed to the internal `VectorBuilder` produced a negative value passed to `new Array[Byte](...)`, throwing `NegativeArraySizeException`. Fixed by guarding the resize with `needed > _tempCapacity` so hints that don't require expansion are silently ignored.

### `copyToArray` returns negative count
All three concrete subclasses (`ByteString1C`, `ByteString1`, `ByteStrings`) computed `dest.length - start` which is negative when `start >= dest.length`, and returned it directly. The Scala collection contract requires a non-negative return. Fixed with `math.max(0, ...)`.

### `ByteString1.apply` factory — negative length bypasses canonical empty
The guard `if (length == 0) empty` did not fire for negative `length` values, so `ByteString1(arr, 0, -5)` silently created a non-canonical empty instance. Fixed to `if (length <= 0) empty`. The defensive `Math.max(0, length)` is retained in the `else` branch for clarity.

---

### `indexOfSlice[B >: Byte]` and `lastIndexOfSlice[B >: Byte]` — unsafe cast
Both methods used `slice.head.asInstanceOf[Byte]` (and `slice(sliceLength-1).asInstanceOf[Byte]`) which throws `ClassCastException` when `B` is not `Byte` at runtime (e.g., `Seq[Int]`). Replaced with a pattern match: when the element is a `Byte`, the optimised `indexOf(byte, ...)` path is taken; otherwise the generic `indexOf[B](elem, ...)` path is used.

---

### `ByteString1.slice` — single allocation
`drop(from).take(until - max(0, from))` created two intermediate `ByteString1` objects. The new implementation computes the clamped range directly and delegates to `ByteString1(bytes, startIndex + lo, hi - lo)` — a single allocation. It also returns `this` when the full range is requested.

### `foreach` overrides in `ByteString1C` and `ByteString1`
The base-class implementation allocated a `ByteIterator`. The new overrides loop over the backing array directly, eliminating the allocation.

### `ByteString.map` — builder-based instead of intermediate collection
The original `fromSpecific(super.map(f))` built an intermediate `IndexedSeq[Byte]`. The new base-class implementation uses `ByteString.newBuilder` directly. `ByteString1C` and `ByteString1` additionally have direct array-based overrides that avoid per-element `apply` calls.

---

## New Tests

- `takeRight` with zero, negative, and over-length arguments for all three concrete types
- `apply(idx)` with out-of-bounds index throws `IndexOutOfBoundsException` on all concrete types
- `copyToArray(xs, start, len)` with `start >= xs.length` returns `0`
- `sizeHint` with negative/small values does not throw and does not corrupt in-progress data
- `grouped(size > length)` yields exactly one group equal to the full ByteString
- `map` correctness for `ByteString1C`, `ByteString1` (with offset), `ByteStrings`, and empty
- `foreach` visits all bytes in order for `ByteString1C`, `ByteString1` (with offset), `ByteStrings`, and empty
- `slice` boundary conditions on `ByteString1`: empty result for inverted/out-of-range, identity when full range, correct clamping
- `indexOfSlice` / `lastIndexOfSlice` with `Seq[Int]` (non-Byte B) — no `ClassCastException`
- `ByteString1.apply` factory returns the canonical `ByteString1.empty` singleton for negative lengths

### CODE REVIEW

**Review by He-Pin — COMMENTED** [review]
## Review: ByteString Bug Fixes, Type Safety & Performance Improvements

Overall this is a solid PR with well-targeted fixes. I've reviewed each change carefully.

### Bug Fixes — Approved

#### 1. `ByteString1C.writeToBuffer` — `copyLength` formula fix (line 418)

**Old:** `Math.min(buffer.remaining, offset + length)`  
**New:** `Math.min(buffer.remaining, length - offset)`

Correct. The old formula was clearly wrong — when `offset > 0`, `offset + length` exceeds the actual remaining bytes from `offset` onwards. For example, with `bytes.length=10`, `offset=5`, `length=10` (bug: length can't exceed backing array, but hypothetically), the old formula would compute `5 + 10 = 15` which is out of bounds.

#### 2. `copyToArray` returns negative count — `math.max(0, ...)` fix (lines 431, 793, 1205)

All three subclasses (`ByteString1C`, `ByteString1`, `ByteStrings`) now guard against `start >= dest.length` producing negative `toCopy`. This is the correct fix — the Scala collection contract requires `copyToArray` to return a non-negative count.

#### 3. `ByteString1.apply` factory — negative length (line 480)

**Old:** `if (length == 0) empty`  
**New:** `if (length <= 0) empty`

Correct. The old code allowed `ByteString1(arr, 0, -5)` to create a non-canonical "empty" instance. The `Math.max(0, ...)` in the else branch is good to keep for defensive clarity.

#### 4. `ByteStringBuilder.sizeHint` — `NegativeArraySizeException` fix (line 2140)

**Old:** `resizeTemp(len - (_length - _tempLength))` unconditionally  
**New:** guards with `if (needed > _tempCapacity)`

This is the right fix. When `sizeHint(n)` is called with `n` smaller than already-committed bytes, `needed` becomes negative and `new Array[Byte](negative)` throws. The guard correctly skips the resize when no expansion is needed.

---

### Type Safety — Approved

#### `indexOfSlice` / `lastIndexOfSlice` — unsafe cast removal (lines 1407-1430, 1486-1527)

The `slice.head.asInstanceOf[Byte]` was a real `ClassCastException` risk when `B` is not `Byte` (e.g., `Seq[Int](97, 98)`). The pattern match approach is correct:

```scala
slice.head match {
  case headByte: Byte => // use optimized indexOf(byte, ...)
  case headElem => // use generic indexOf[B](elem, ...)
}
```

One minor observation: the `case headElem` branch uses `indexOf(headElem, pos)` and `lastIndexOf(headElem, currEnd)` which rely on `equals` semantics. For `Seq[Int]` with `Byte` values, `97.toByte == 97` is `true` in Scala, so this works correctly. 

---

### Performance Improvements — Approved with minor notes

#### 1. `ByteString1.slice` — single allocation (lines 544-551)

The new implementation avoids two intermediate allocations (`drop` then `take`) and directly computes the clamped range. The identity check `if (lo == 0 && hi == length) this` is a nice touch for the common case.

#### 2. `foreach` overrides (lines 442-448, 810-817)

Eliminating `ByteIterator` allocation in `foreach` is a good optimization. The while-loop over the backing array is straightforward and efficient.

#### 3. `map` overrides (lines 450-459, 819-828, 1364-1373)

All three levels now avoid intermediate `IndexedSeq[Byte]` allocation:
- `ByteString` base class uses `ByteString.newBuilder`
- `ByteString1C` and `ByteString1` have direct array-based implementations

**Note:** The base-class `map` on line 1364 uses `b += f(apply(i))` which goes through `apply(i)` — this is O(1) for single-segment types but O(n) for `ByteStrings`. For the builder-based approach this is fine since the builder already amortizes the cost, but it's worth noting that `ByteStrings.map` doesn't get a specialized array-based override like the other two types.

---

### Tests — Approved

The test coverage is comprehensive and targets all the edge cases:

- ✅ `takeRight` boundary values (zero, negative, over-length)
- ✅ `apply(idx)` out-of-bounds throws `IndexOutOfBoundsException`
- ✅ `copyToArray` returns 0 when `start >= dest.length`
- ✅ `sizeHint` with negative/small values doesn't throw
- ✅ `grouped(size > length)` yields exactly one group
- ✅ `map` correctness on all concrete types
- ✅ `foreach` visits all bytes in order
- ✅ `slice` boundary conditions on `ByteString1`
- ✅ `indexOfSlice` / `lastIndexOfSlice` with `Seq[Int]` — no `ClassCastException`
- ✅ `ByteString1.apply` factory returns canonical empty for non-positive length

One suggestion: consider adding a test for `writeToBuffer` with `offset > 0` to verify the `copyLength` fix doesn't regress in the future.

---

### Style / Formatting

The `span` formatting change (line 1316) is purely stylistic — breaking the one-liner into multiple lines. This is fine and improves readability.

---

### Summary

This PR fixes genuine bugs (some potentially severe like `ArrayIndexOutOfBoundsException` and `NegativeArraySizeException`), improves type safety, and adds meaningful performance optimizations. The test coverage is thorough.

**✅ Approved — Ready to merge.**

**Review by He-Pin — APPROVED** [review]
## Review: ByteString Bug Fixes, Type Safety & Performance Improvements

### Bug Fixes — All Correct

1. **`ByteString1C.writeToBuffer` (line 418):** `Math.min(buffer.remaining, offset + length)` → `Math.min(buffer.remaining, length - offset)`. Correct — old formula could exceed array bounds when `offset > 0`.

2. **`copyToArray` negative return (lines 431, 793, 1205):** Added `math.max(0, ...)` guard in all three subclasses. Correct — Scala collection contract requires non-negative return.

3. **`ByteString1.apply` negative length (line 480):** `if (length == 0)` → `if (length <= 0)`. Correct — prevents non-canonical empty instances.

4. **`ByteStringBuilder.sizeHint` (line 2140):** Guard with `if (needed > _tempCapacity)`. Correct — prevents `NegativeArraySizeException` when hint is smaller than already-committed bytes.

### Type Safety — Correct

`indexOfSlice` / `lastIndexOfSlice` unsafe cast replaced with pattern match. The `slice.head.asInstanceOf[Byte]` was a real `ClassCastException` risk with `Seq[Int]` etc. Pattern match correctly dispatches to optimized path for `Byte` or generic path for other `B >: Byte`.

### Performance — Well Done

- `slice` avoids two intermediate allocations, computes clamped range directly
- `foreach` overrides eliminate `ByteIterator` allocation
- `map` overrides avoid intermediate `IndexedSeq[Byte]` via builder-based approach

### Tests — Comprehensive

Covers all edge cases: boundary values, negative inputs, out-of-bounds access, type safety with `Seq[Int]`.

**✅ Approved.**

File: actor/src/main/scala/org/apache/pekko/util/ByteString.scala [hunk 1]
```diff
@@ -415,7 +415,7 @@ object ByteString {
 
     /** INTERNAL API: Specialized for internal use, writing multiple ByteString1C into the same ByteBuffer. */
     private[pekko] def writeToBuffer(buffer: ByteBuffer, offset: Int): Int = {
-      val copyLength = Math.min(buffer.remaining, offset + length)
+      val copyLength = Math.min(buffer.remaining, length - offset)
```
Comment (He-Pin) [inline actor/src/main/scala/org/apache/pekko/util/ByteString.scala:418]: wow,nice find

**Review by He-Pin — APPROVED** [review]
lgtm
